### PR TITLE
SCP-3367: PAB CLI for easing the starting of PAB + components

### PR DESCRIPTION
**Pull request by koslambrou**
PAB CLI for easing the starting of components based on supported scenarios/use-cases.

Provides a `pab-cli` command line app in the `nix-shell`. This CLI can start the PAB in the currently supported scenarios (with mocked components, real components on mainnet or testnet, and optionally the cardano-wallet).

Also provides the `plutus-chain-index` executable in the `nix-shell`.

Next step is to integrate this tool in `plutus-starter`.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-pab-executables/pab-cli/CommandParser.hs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+              AppOpts (AppOpts), ChainIndexOpts (ChainIndexOpts), ChainIndexPort (ChainIndexPort),
+              ConfigCommand (MockNetCommand, NodeRemoteWalletCommand, NodeWBECommand), NetworkName (Mainnet, Testnet),
+              NodeDirectory (NodeDirectory), NodeOpts (NodeOpts), NodePort (NodePort), PABDirectory (PABDirectory),
+              PABExe (PABExe), PABOpts (PABOpts), PABPort (PABPort), WalletPort (WalletPort))
+
+patterns :: Docopt
+patterns = [docopt|
+PAB CLI. This script allows the user to run the PAB in a hosted scenario, and provides good defaults for common usecases.
+
+THIS IS AN EXPERIMENT! DO NOT USE IN A PRODUCTION ENVIRONNEMENT.
```
Comment (james-iohk) [inline plutus-pab-executables/pab-cli/CommandParser.hs:30]: "ENVIRONMENT"
:)

Comment (koslambrou) [reply plutus-pab-executables/pab-cli/CommandParser.hs:30]: Oups, the french got the better of me.

File: plutus-pab-executables/pab-cli/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                                          [default:9082]
+    --wallet-port <port>                  wallet server port number
+                                          [default:9081]
+```
+
+The tool currently supports the following three scenarios:
+
+1- You want to run your PAB webserver with mocked components (node, wallet and chain-index).
+
+2- You want to run the cardano-wallet backend, cardano node, chain-index and (optionnaly) your PAB webserver on the Cardano mainnet or testnet.
```
Comment (james-iohk) [inline plutus-pab-executables/pab-cli/README.md:52]: "optionnaly" -> "optionally" (same in point 3 below)

File: plutus-pab-executables/pab-cli/App.hs [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+-- | Run list of actions asynchroniously and wait until any one of them finishes.
+runAsyncActions :: [IO ()] -> IO ()
+-- runAsyncActions ios = do
+--     asyncs <- mapM (\io -> withAsync io pure) ios
+--     print $ length asyncs
+--     void $ waitAny asyncs
+runAsyncActions ios = void $ go ios []
+    where go [] asyncs = waitAny asyncs
+          go (x:xs) asyncs = do
+              withAsync x $ \a -> go xs (a : asyncs)
```
Comment (sjoerdvisscher) [inline plutus-pab-executables/pab-cli/App.hs:225]: The `Alternative` instance of `Concurrently` does `race` which I think is what you want here, so:
```haskell
runAsyncActions = ala (Alt . Concurrently) foldMap
```

File: plutus-pab-executables/pab-cli/CommandParser.hs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+import System.Directory (Permissions (executable), doesFileExist, getPermissions)
+import System.FilePath ((</>))
+import Text.Read (readMaybe)
+import Types (AppError (ChainIndexPortError, NodePortError, PABDBPoolSizeError, PABExeNotExecutableError, PABExeNotProvidedError, PABPortError, PABResumeFromBlockIdNotSpecifiedError, PABResumeFromSlotNotANumberError, PABResumeFromSlotNotSpecifiedError, PABRollbackHistoryNotNumberError, UnspecifiedNodeNetworkError, WalletPortError),
+              AppOpts (AppOpts), ChainIndexOpts (ChainIndexOpts), ChainIndexPort (ChainIndexPort),
+              ConfigCommand (MockNetCommand, NodeRemoteWalletCommand, NodeWBECommand), NetworkName (Mainnet, Testnet),
+              NodeDirectory (NodeDirectory), NodeOpts (NodeOpts), NodePort (NodePort), PABDirectory (PABDirectory),
+              PABExe (PABExe), PABOpts (PABOpts), PABPort (PABPort), WalletPort (WalletPort))
+
+patterns :: Docopt
```
Comment (sjoerdvisscher) [inline plutus-pab-executables/pab-cli/CommandParser.hs:26]: Add a comment here that if this is changed, the README should also be updated.

**Review by sjoerdvisscher — APPROVED** [review]

File: plutus-pab-executables/pab-cli/App.hs (hunk 3 above)
Comment (koslambrou) [reply plutus-pab-executables/pab-cli/App.hs]: Nice!

**Pull request merged**